design-documents/graph-ql/path-validator.md

```diff
@@ -0,0 +1,100 @@
+
+## Overview
+### Current situation
+In the Magento 2 framework we can only validate URLs `http://www.somedomain.com/some/path`.
+We don't have the possibility to configure this validator to test relative urls, paths as Zend framework refers to them.
+We should have such kind of validator available for cases where we would just need to validate a relative path and also
+eliminate the directory traversal when such an input comes from the user. 
+
+### Problem
+Such situation of validating a url route did occur with the proposal
+[Domain Whitelist for Configurable 3rd Party Redirects](https://github.com/magento/architecture/pull/204).
+We just don't have a service class to validate url routes, for code re-usage, with that can be injected in all resolvers or
+classes that need this validator such as PayPal Express and others.
+
+### Proposed solution
+As this validator doesn't have anything to do with GraphQl, the best place for it is to add a new validator in framework
+similar to existing `\Magento\Framework\Url\Validator`. So this new class in framework will need approval.
+
+Looking at the implementation of the `\Magento\Framework\Url\Validator` it's based on `\Zend_Validate_Abstract` so it
+complies with our standard of not using native php functions such as parse `parse_url`. Nor we would want to use regex.
+
+
+The new class `RouteValidator` will look like the following and it's namespace will be `Magento\Framework\Url`:
+```   
+   namespace Magento\Framework\Url;
+   
+   /**
+    * Validate url routes
+    */
+   class RouteValidator extends \Zend_Validate_Abstract
+   {
+       /**
+        * @var \Zend\Validator\Uri
+        */
+       private $validator;
+   
+       /**
+        * @param \Zend\Validator\Uri $validator
+        */
+       public function __construct(\Zend\Validator\Uri $validator)
+       {
+           // set translated message template
+           $this->setMessage((string)new \Magento\Framework\Phrase("Invalid URL '%value%'."), Validator::INVALID_URL);
+           $this->validator = $validator;
+       }
+   
+       /**
+        * Validation failure message template definitions
+        *
+        * @var array
+        */
+       protected $_messageTemplates = [Validator::INVALID_URL => "Invalid URL '%value%'."];
+   
+       /**
+        * Validate url route
+        *
+        * @param string $value
+        * @return bool
+        */
+       public function isValid($value)
+       {
+           // configure the validator
+           $this->validator->setAllowRelative(true);
+           $this->validator->setAllowAbsolute(false);
+           $this->_setValue($value);
+   
+           $valid = $this->validator->isValid($value)
+               && $this->validator->getUriHandler()->getQuery() === null
+                 //prevent directory traversal
+               && strpos($value, '..') === false;
+   
+           if (!$valid) {
+               $this->_error(Validator::INVALID_URL);
```

Comment:
I think it doesn't make sense to add some errors because in general, the method's call result should be if the URL is valid or not. So there is no reason to have such methods like `getMessages` with errors.